### PR TITLE
Remove replay protection from the auth crate

### LINKED ISSUE #590: Auth: Remove nonce functionality from soroban-auth
I think we should remove the nonce functionality from soroban-auth.

By bundling nonce validation into the auth interface it feels like we're making the interface handle two unrelated concerns, and that both constrains where the auth interface can be used, and the cognitive load in learning how it works and what it does. That last concern is a pretty big deal because we need developers to understand how auth works, and what it does, and when it is doing multiple things it is harder to understand the subtleties necessary to know if you are using it correctly.

Also, looking at [EIP712](https://eips.ethereum.org/EIPS/eip-712) it is interesting to note that it also explicitly calls out the importance that its scheme does not prevent replay and how it is the applications responsibility to do that.

cc @tamirms @sisuresh @jonjove

---

>+1 What I find most interesting about 712 is two things:
- the domain is very simple and doesn't include a contracts full call stack. Is there anywhere in these sta dards that a full stack shows up?
- the domain includes the verifying contract ID. Afaict this isn't necessarily the receiving contract? Is this accurate?
- the nonce is absent and it is up to the application to prevent replay, which I like and I think it would simplify our auth interface to consider this.

_Originally posted by @leighmcculloch in https://github.com/stellar/rs-soroban-sdk/issues/577#issuecomment-1243923019_

### PR DESCRIPTION
**Pull request by sisuresh**
### What
Resolves #590, #533, and #534.

### CODE REVIEW

File: soroban-auth/tests/test.rs [hunk 1]
```diff
@@ -14,31 +13,34 @@ pub enum DataKey {
...
-        let nonce = Self::read_nonce(e, id);
-        e.contract_data().set(key, &nonce + 1);
-        nonce
-    }
+    let key = DataKey::Nonce(id.clone());
+    let nonce = read_nonce(e, id);
 
-    fn signature(&self) -> &Signature {
-        &self.0
+    if nonce != *expected_nonce {
```
Comment (leighmcculloch) [inline soroban-auth/tests/test.rs:40]: Does this equality check not work without the `*`? I think I fixed that recently, it should work now.

**Review by leighmcculloch — APPROVED** [review]

File: soroban-auth/tests/test.rs (hunk 1 above)
Comment (sisuresh) [reply soroban-auth/tests/test.rs:40]: No I get this error
```
can't compare `soroban_sdk::BigInt` with `&soroban_sdk::BigInt`
the trait `PartialEq<&soroban_sdk::BigInt>` is not implemented for `soroban_sdk::BigInt`
```

Comment (leighmcculloch) [reply soroban-auth/tests/test.rs:40]: Ok no worries. I'll fix it, but no need to do anything differently than you are here.

Comment (leighmcculloch) [reply soroban-auth/tests/test.rs:40]: Fixing in https://github.com/stellar/rs-soroban-sdk/pull/598.

Comment (sisuresh) [reply soroban-auth/tests/test.rs]: updated

**Pull request merged**